### PR TITLE
Instrumental strategy pablo/adding collective pallet

### PR DESCRIPTION
**Pull request by belousm**
## Issue
[*2rb3egb*](https://app.clickup.com/t/2rb3egb)

## Description
*Remove the custom addition of access rights and privileged users and replace it with a `pallet_collective` functionality.*

### CODE REVIEW

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-2rb3egb Add extrensic "set_members" call from palette_collective to "set_access"](https://app.clickup.com/t/2rb3egb)

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-2rb3eaj Integrating pallet_collective with instrumental](https://app.clickup.com/t/2rb3eaj)

**Review by saruman9 — CHANGES_REQUESTED** [review]
When I try to execute `cargo check --all-targets --all-features` in `instrumental{-strategy{-pablo,},}` directories I get:
```
error[E0046]: not all trait items implemented, missing: `successful_origin`
    --> /home/neo/.cargo/git/checkouts/substrate-01ceebf642695c01/616d33e/frame/collective/src/lib.rs:986:1
     |
986  | / impl<
987  | |         O: Into<Result<RawOrigin<AccountId, I>, O>> + From<RawOrigin<AccountId, I>>,
988  | |         I,
989  | |         AccountId: Decode,
...    |
1006 | |     }
1007 | | }
     | |_^ missing `successful_origin` in implementation
     |
     = help: implement the missing item: `fn successful_origin() -> OuterOrigin { todo!() }`

error[E0046]: not all trait items implemented, missing: `successful_origin`
    --> /home/neo/.cargo/git/checkouts/substrate-01ceebf642695c01/616d33e/frame/collective/src/lib.rs:1010:1
     |
1010 | / impl<
1011 | |         O: Into<Result<RawOrigin<AccountId, I>, O>> + From<RawOrigin<AccountId, I>>,
1012 | |         AccountId,
1013 | |         I,
...    |
1028 | |     }
1029 | | }
     | |_^ missing `successful_origin` in implementation
     |
     = help: implement the missing item: `fn successful_origin() -> OuterOrigin { todo!() }`

error[E0046]: not all trait items implemented, missing: `successful_origin`
    --> /home/neo/.cargo/git/checkouts/substrate-01ceebf642695c01/616d33e/frame/collective/src/lib.rs:1034:1
     |
1034 | / impl<
1035 | |         O: Into<Result<RawOrigin<AccountId, I>, O>> + From<RawOrigin<AccountId, I>>,
1036 | |         AccountId,
1037 | |         I,
...    |
1053 | |     }
1054 | | }
     | |_^ missing `successful_origin` in implementation
     |
     = help: implement the missing item: `fn successful_origin() -> OuterOrigin { todo!() }`

error[E0046]: not all trait items implemented, missing: `successful_origin`
    --> /home/neo/.cargo/git/checkouts/substrate-01ceebf642695c01/616d33e/frame/collective/src/lib.rs:1059:1
     |
1059 | / impl<
1060 | |         O: Into<Result<RawOrigin<AccountId, I>, O>> + From<RawOrigin<AccountId, I>>,
1061 | |         AccountId,
1062 | |         I,
...    |
1078 | |     }
1079 | | }
     | |_^ missing `successful_origin` in implementation
     |
     = help: implement the missing item: `fn successful_origin() -> OuterOrigin { todo!() }`
 ```
 
I think the problem arises when using `pallet-collective` without `std` feature. Could you research this, please? Because there is no such problem in other pallets of Composable.

File: frame/instrumental/src/mock/runtime.rs [hunk 1]
```diff
@@ -319,6 +347,7 @@ frame_support::construct_runtime!(
 		Assets: pallet_assets::{Pallet, Call, Storage},
 		GovernanceRegistry: pallet_governance_registry::{Pallet, Call, Storage, Event<T>},
 		Timestamp: pallet_timestamp::{Pallet, Call, Storage},
+		CollectiveInstrumental: pallet_collective::<Instance1>::{Pallet, Call, Event<T>, Origin<T>, Config<T>},
```
Comment (saruman9) [inline frame/instrumental/src/mock/runtime.rs:350]: You created the type (`type InstrumentalPabloCollective = pallet_collective::Instance1;`) above, but here you use `Instance1`. Maybe `InstrumentalPabloCollective` will be more suitable here. What do you think?

File: frame/instrumental-strategy/src/mock/runtime.rs [hunk 2]
```diff
@@ -19,7 +19,14 @@ pub type PoolId = u128;
 pub type Moment = composable_traits::time::Timestamp;
 pub type VaultId = u64;
 
-pub const MILLISECS_PER_BLOCK: u64 = 12000;
+const NATIVE_ASSET: CurrencyId = CurrencyId::PICA;
```
Comment (saruman9) [inline frame/instrumental-strategy/src/mock/runtime.rs:22]:
```
warning: constant is never used: `NATIVE_ASSET`
```

File: frame/instrumental-strategy-pablo/src/tests.rs [hunk 3]
```diff
@@ -75,25 +78,93 @@ mod rebalance {
...
-				Origin::signed(ADMIN),
-				base_asset,
-				pool_id
+			let members_count: MemberCount = 5;
+			CollectiveInstrumental::set_members(
+				Origin::root(),
+				vec![ADMIN, ALICE, BOB],
+				None,
+				members_count,
+			);
```
[the remark this answers is not in the document]
Comment (belousm) [reply frame/instrumental-strategy-pablo/src/tests.rs]: Thanks for your attention! Fixed!

File: frame/instrumental-strategy/src/mock/runtime.rs (hunk 2 above)
Comment (belousm) [reply frame/instrumental-strategy/src/mock/runtime.rs:22]: [It's from `instrumental` branch ](https://github.com/ComposableFi/composable/blob/f9be10880770ceea4460f44e2ca6ac7ace46139d/frame/instrumental-strategy-pablo/src/mock/runtime.rs#L28)

File: frame/instrumental/src/mock/runtime.rs (hunk 1 above)
Comment (belousm) [reply frame/instrumental/src/mock/runtime.rs:350]: I can't use in `construct_runtime!` the same name which was used for type annotation.

Comment (belousm) [reply frame/instrumental/src/mock/runtime.rs:350]: I can rename it something like `CollectivePabloStrategyInstrumental`

File: frame/instrumental-strategy-pablo/src/lib.rs [hunk 4]
```diff
@@ -246,12 +242,7 @@ pub mod pallet {
 			asset_id: T::AssetId,
 			pool_id: T::PoolId,
 		) -> DispatchResultWithPostInfo {
-			let who = ensure_signed(origin)?;
-			let access_right = Self::admin_accounts(who).ok_or(Error::<T>::NotAdminAccount)?;
-			ensure!(
-				access_right == AccessRights::Full || access_right == AccessRights::SetPoolId,
-				Error::<T>::NotEnoughAccessRights
-			);
+			T::ExternalOrigin::ensure_origin(origin)?;
```
Comment (saruman9) [inline frame/instrumental-strategy-pablo/src/lib.rs:245]: GitHub doesn't allow to create comments for code that has not been changed (strange behavior I think). I think these errors are no longer needed:
- `NotEnoughAccessRights`
- `NotAdminAccount`

Comment (belousm) [reply frame/instrumental-strategy-pablo/src/lib.rs:241]: Sure!

Comment (belousm) [reply frame/instrumental-strategy-pablo/src/lib.rs:241]: Sure!

File: frame/instrumental-strategy/src/mock/runtime.rs (hunk 2 above)
Comment (belousm) [reply frame/instrumental-strategy/src/mock/runtime.rs]: Resolved

File: frame/instrumental-strategy-pablo/src/tests.rs [hunk 5]
```diff
@@ -75,25 +78,95 @@ mod rebalance {
...
-				base_asset,
-				pool_id
+			let members_count: MemberCount = 5;
+			CollectiveInstrumental::set_members(
+				Origin::root(),
+				vec![ADMIN, ALICE, BOB],
+				None,
+				members_count,
+			)
+			.ok();
```
Comment (saruman9) [inline frame/instrumental-strategy-pablo/src/tests.rs:90]: You already used `assert_ok!` macro everywhere, please use it here too.

File: frame/instrumental-strategy-pablo/src/tests.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (saruman9) [reply frame/instrumental-strategy-pablo/src/tests.rs]: https://github.com/ComposableFi/composable/pull/1333#discussion_r923239554

File: frame/instrumental-strategy-pablo/src/tests.rs [hunk 6]
```diff
@@ -75,25 +78,95 @@ mod rebalance {
...
+				asset_id: base_asset,
+				pool_id,
+			});
+			let proposal_len: u32 = proposal.using_encoded(|p| p.len() as u32);
+			let proposal_weight = proposal.get_dispatch_info().weight;
+			let hash: H256 = BlakeTwo256::hash_of(&proposal);
+			assert_ok!(CollectiveInstrumental::propose(
+				Origin::signed(ALICE),
+				2,
+				Box::new(proposal.clone()),
```
Comment (saruman9) [inline frame/instrumental-strategy-pablo/src/tests.rs:156]: Redundant clone.

File: frame/instrumental-strategy-pablo/Cargo.toml [hunk 7]
```diff
@@ -38,7 +38,8 @@ pallet-currency-factory = { path = "../currency-factory", default-features = fal
 pallet-governance-registry = { path = "../governance-registry", default-features = false }
 pallet-instrumental = { version = "0.1.0", path = "../instrumental", default-features = false }
 pallet-pablo = { path = "../pablo", default-features = false }
-pallet-timestamp = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.24", default-features = false }
+pallet-timestamp = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.22", default-features = false }
+pallet-collective = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.22", default-features = false }
```
Comment (saruman9) [inline frame/instrumental-strategy-pablo/Cargo.toml:42]: Old versions are used (current 0.9.24) here and in other `Cargo.toml` files.

File: frame/instrumental-strategy-pablo/src/tests.rs [hunk 8]
```diff
@@ -75,25 +78,95 @@ mod rebalance {
...
+					occured_associated_pool_with_asset = true;
+				}
+			});
+			assert!(occured_associated_pool_with_asset);
+		});
+	}
 
-			assert_ok!(PabloStrategy::associate_vault(&vault_id));
-
-			assert_ok!(PabloStrategy::rebalance());
```
[the remark this answers is not in the document]
Comment (belousm) [reply frame/instrumental-strategy-pablo/src/tests.rs:92]: So, now, as you [mentioned](https://app.clickup.com/t/2m7v9u0), we should redesign  `rebalance` function. So, I decided to delete all `rebalance` tests and added tests for `set pool_id for asset_id` function.

[the remark this answers is not in the document]
Comment (saruman9) [reply frame/instrumental-strategy-pablo/src/tests.rs:92]: Yes, this is correct, but tests will not change drastically. Could you return `rebalance`'s tests, please?

File: frame/instrumental-strategy/src/mock/runtime.rs [hunk 9]
```diff
@@ -1,16 +1,16 @@
+use crate as pallet_instrumental_strategy;
 use frame_support::{parameter_types, traits::Everything, PalletId};
 use frame_system::{EnsureRoot, EnsureSigned};
 use orml_traits::parameter_type_with_key;
-use primitives::currency::{CurrencyId, ValidateCurrencyId};
+use pallet_collective::EnsureProportionAtLeast;
+use primitives::currency::CurrencyId;
```
Comment (saruman9) [inline frame/instrumental-strategy/src/mock/runtime.rs:6]: You accidentally deleted `ValidateCurrencyId` import. Could you return it, please?
> error[E0412]: cannot find type `ValidateCurrencyId` in this scope

**Comment by belousm** [comment]
> When I try to execute `cargo check --all-targets --all-features` in `instrumental{-strategy{-pablo,},}` directories I get:
> 
> ```
> error[E0046]: not all trait items implemented, missing: `successful_origin`
>     --> /home/neo/.cargo/git/checkouts/substrate-01ceebf642695c01/616d33e/frame/collective/src/lib.rs:986:1
>      |
> 986  | / impl<
> 987  | |         O: Into<Result<RawOrigin<AccountId, I>, O>> + From<RawOrigin<AccountId, I>>,
> 988  | |         I,
> 989  | |         AccountId: Decode,
> ...    |
> 1006 | |     }
> 1007 | | }
>      | |_^ missing `successful_origin` in implementation
>      |
>      = help: implement the missing item: `fn successful_origin() -> OuterOrigin { todo!() }`
> 
> error[E0046]: not all trait items implemented, missing: `successful_origin`
>     --> /home/neo/.cargo/git/checkouts/substrate-01ceebf642695c01/616d33e/frame/collective/src/lib.rs:1010:1
>      |
> 1010 | / impl<
> 1011 | |         O: Into<Result<RawOrigin<AccountId, I>, O>> + From<RawOrigin<AccountId, I>>,
> 1012 | |         AccountId,
> 1013 | |         I,
> ...    |
> 1028 | |     }
> 1029 | | }
>      | |_^ missing `successful_origin` in implementation
>      |
>      = help: implement the missing item: `fn successful_origin() -> OuterOrigin { todo!() }`
> 
> error[E0046]: not all trait items implemented, missing: `successful_origin`
>     --> /home/neo/.cargo/git/checkouts/substrate-01ceebf642695c01/616d33e/frame/collective/src/lib.rs:1034:1
>      |
> 1034 | / impl<
> 1035 | |         O: Into<Result<RawOrigin<AccountId, I>, O>> + From<RawOrigin<AccountId, I>>,
> 1036 | |         AccountId,
> 1037 | |         I,
> ...    |
> 1053 | |     }
> 1054 | | }
>      | |_^ missing `successful_origin` in implementation
>      |
>      = help: implement the missing item: `fn successful_origin() -> OuterOrigin { todo!() }`
> 
> error[E0046]: not all trait items implemented, missing: `successful_origin`
>     --> /home/neo/.cargo/git/checkouts/substrate-01ceebf642695c01/616d33e/frame/collective/src/lib.rs:1059:1
>      |
> 1059 | / impl<
> 1060 | |         O: Into<Result<RawOrigin<AccountId, I>, O>> + From<RawOrigin<AccountId, I>>,
> 1061 | |         AccountId,
> 1062 | |         I,
> ...    |
> 1078 | |     }
> 1079 | | }
>      | |_^ missing `successful_origin` in implementation
>      |
>      = help: implement the missing item: `fn successful_origin() -> OuterOrigin { todo!() }`
> ```
> 
> I think the problem arises when using `pallet-collective` without `std` feature. Could you research this, please? Because there is no such problem in other pallets of Composable.

Resolved!

File: frame/instrumental-strategy-pablo/src/tests.rs [hunk 10]
```diff
@@ -97,3 +115,110 @@ mod rebalance {
...
+			system_events.iter().for_each(|event_record| {
+				if event_record.event ==
+					Event::PabloStrategy(pallet::Event::AssociatedPoolWithAsset {
+						asset_id: base_asset,
+						pool_id,
+					}) {
+					occured_associated_pool_with_asset = true;
+				}
+			});
+			assert!(occured_associated_pool_with_asset);
```
Comment (saruman9) [inline frame/instrumental-strategy-pablo/src/tests.rs:177]:
```suggestion
			System::assert_has_event(Event::PabloStrategy(
				pallet::Event::AssociatedPoolWithAsset { asset_id: base_asset, pool_id },
			));
```

**Review by saruman9 — CHANGES_REQUESTED** [review]
@belousm, could you syncronize with `main` branch, please? We agreed on this [here](https://app.clickup.com/t/2m24e1f?comment=900859127).

**Review by saruman9 — APPROVED** [review]
https://github.com/ComposableFi/composable/pull/1333#pullrequestreview-1042759203

**Review by kevin-composable — APPROVED** [review]
Everything looks good, I would just consider creating helper functionality for initializing collectives as that will be duplicate code in most tests

**Pull request merged**